app/services/spl/validate_card_service.rb

```diff
@@ -0,0 +1,111 @@
+      true
+    rescue StandardError
+      @user.save
+      raise
+    end
+
+    private
+
+    def check_for_errors(response_body)
+      raise SplCardValidationError, response_body.dig('msg') if response_body["errorCode"] != "0"
+
+      card_assignment = cards_assigned_user(@card_number)
+
+      if card_assigned_to_different_user(card_assignment)
+        raise SplCardValidationError, I18n.t("spl.card_validation.errors.wrong_owner")
+      end
+      raise SplCardValidationError, I18n.t("spl.card_validation.errors.card_not_match") if user_have_different_card
+      raise SplCardValidationError, I18n.t("spl.card_validation.errors.card_not_active") if response_body
+                                                                                       .dig("response",
+                                                                                            "card", "status") != "A"
+    end
+
+    def verify_card_request
+      url = URI.parse(ENV["SPL_CHECK_CARD_URL"])
+      http = Net::HTTP.new(url.host, url.port)
+      http.use_ssl = true
+
+      request = Net::HTTP::Post.new(url)
+      request["Content-Type"] = "application/json"
+
+      request.body = body(@card_number, DateTime.current).to_json
+
+      response = http.request(request)
+      response.body
+    end
+
+    def body(card_number, date) # rubocop: disable Metrics/MethodLength
+      date_in_ms = date.to_i * 1000
+      uuid = SecureRandom.uuid
+      {
+        ver: 4,
+        requestId: uuid,
+        apiUser: ENV["SPL_API_USER"],
+        apiToken: ENV["SPL_API_TOKEN"],
+        partnerCode: ENV["SPL_PARTNER_CODE"],
+        placeCode: ENV["SPL_PLACE_CODE"],
+        date: date_in_ms,
+        cardNo: card_number,
+        signature: signature(date_in_ms, card_number),
+        extendedPersonalInfo: true
+      }
+    end
+
+    def signature(date, card_number)
+      data = "#{ENV["SPL_PARTNER_CODE"]}#{ENV["SPL_PLACE_CODE"]}#{date}#{card_number}"
+      Rails.logger.debug data.inspect
+      signature_base = Digest::SHA256.hexdigest(data)
+      Digest::SHA256.hexdigest(signature_base + ENV["SPL_POS_KEY"])
+    end
+
+    def cards_assigned_user(card_number)
+      Spree::User.find { |u| u.public_metadata["spl_no_card"] == card_number }
+    end
+
+    def card_assigned_to_different_user(card_assignment)
+      card_assignment && card_assignment.id != @user.id
+    end
+
+    def user_have_different_card
+      @user.public_metadata["spl_no_card"] && @user.public_metadata["spl_no_card"] != @card_number
+    end
+
+    def assign_card_to_user
+      if @user.public_metadata["spl_no_card"] == @card_number && @user.public_metadata["spl_card_active"].present?
+        return
+      end
+
+      @user.public_metadata = @user.public_metadata.merge(
+        {
+          "spl_no_card" => @card_number,
+          "spl_card_active" => true
+        }
+      )
+      current_order = @user.orders.last
+      current_order.public_metadata[:spl_card_active] = true
```

Comment:
Z ważnych rzeczy tu jeszcze trzeba się upewnić, że robisz update dla zamówienia, które jest w trakcie robienia. Poponowałabym po [statusie](https://spreecommerce.org/docs/developer/core-concepts/orders#the-order-state-machine). Możemy się upewnić w poniedziałek czy pojawia się status `confirm` - bo takiego nie kojarzę - ale tu na pewno przydało by się coś w stylu:
```suggestion
      current_order = @user.orders.last
      current_order.update(public_metadata: public_metadata.merge('spl_card_active' => true, 'spl_no_card' => @card_number) if ["cart", "address", "delivery", "payment"].include?(current_order.state)
```
Tu przy okazji pozbywamy się tej linijki 108 z save.

W takim przypadku można też wynieść hasz:
`{
          "spl_no_card" => @card_number,
          "spl_card_active" => true
 }` do zmiennej i użyć go przy mergu public metadata dla usera jak i orderu